File: backend/src/main/java/com/coing/global/config/SecurityConfig.java
package com.coing.global.config;

import java.util.Arrays;

import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.config.http.SessionCreationPolicy;
import org.springframework.security.crypto.argon2.Argon2PasswordEncoder;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.security.web.SecurityFilterChain;
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter;
import org.springframework.web.cors.CorsConfiguration;
import org.springframework.web.cors.CorsConfigurationSource;
import org.springframework.web.cors.UrlBasedCorsConfigurationSource;

import com.coing.global.security.CustomAuthenticationEntryPoint;
import com.coing.global.security.filter.JwtAuthenticationFilter;

import lombok.RequiredArgsConstructor;

@Configuration
@RequiredArgsConstructor
public class SecurityConfig {

	private final JwtAuthenticationFilter jwtAuthenticationFilter;
	private final CustomAuthenticationEntryPoint customAuthenticationEntryPoint;

	@Bean
	public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Exception {
		http
			// CORS 설정 추가
			.cors().and()
			// CSRF 보호 비활성화
			.csrf(csrf -> csrf.disable())
			// H2 콘솔 접근을 허용하기 위해 frameOptions 비활성화
			.headers(headers -> headers.frameOptions(frame -> frame.disable()))
			// 세션을 사용하지 않는 stateless 설정
			.sessionManagement(session -> session.sessionCreationPolicy(SessionCreationPolicy.STATELESS))
			// 모든 요청에 대해 permitAll()
			.authorizeHttpRequests(authz -> authz.anyRequest().permitAll())
			// 커스텀 AuthenticationEntryPoint 및 JwtAuthenticationFilter 추가
			.exceptionHandling(exception -> exception.authenticationEntryPoint(customAuthenticationEntryPoint))
			.addFilterBefore(jwtAuthenticationFilter, UsernamePasswordAuthenticationFilter.class);

		return http.build();
	}

	@Bean
	public CorsConfigurationSource corsConfigurationSource() {
		CorsConfiguration configuration = new CorsConfiguration();
		// 모든 출처 허용 (운영 환경 배포 시 수정 필요)
		configuration.setAllowedOriginPatterns(Arrays.asList("*"));
		configuration.setAllowedMethods(Arrays.asList("GET", "POST", "PUT", "DELETE", "OPTIONS"));
		configuration.setAllowedHeaders(Arrays.asList("*"));
		configuration.setAllowCredentials(true);
		configuration.addExposedHeader("Authorization");
		UrlBasedCorsConfigurationSource source = new UrlBasedCorsConfigurationSource();
		// 모든 엔드포인트에 대해 CORS 설정 적용
		source.registerCorsConfiguration("/**", configuration);
		return source;
	}

	@Bean
	public PasswordEncoder passwordEncoder() {
		// 매개변수: saltLength, hashLength, parallelism, memory, iterations
		return new Argon2PasswordEncoder(16, 32, 1, 4096, 3);
	}
}

File: backend/src/main/java/com/coing/global/config/SecurityConfig.kt
package com.coing.global.config

import com.coing.domain.user.service.OAuth2UserService
import com.coing.global.security.CustomAuthenticationEntryPoint
import com.coing.global.security.filter.JwtAuthenticationFilter
import com.coing.global.security.handler.OAuth2LoginFailureHandler
import com.coing.global.security.handler.OAuth2LoginSuccessHandler
import org.springframework.beans.factory.annotation.Value
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.security.config.annotation.web.builders.HttpSecurity
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity
import org.springframework.security.config.annotation.web.configurers.AbstractHttpConfigurer
import org.springframework.security.config.http.SessionCreationPolicy
import org.springframework.security.crypto.argon2.Argon2PasswordEncoder
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.security.web.SecurityFilterChain
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter
import org.springframework.web.cors.CorsConfiguration
import org.springframework.web.cors.CorsConfigurationSource
import org.springframework.web.cors.UrlBasedCorsConfigurationSource

@Configuration
@EnableWebSecurity
class SecurityConfig(
	private val jwtAuthenticationFilter: JwtAuthenticationFilter,
	private val customAuthenticationEntryPoint: CustomAuthenticationEntryPoint,
	private val oAuth2UserService: OAuth2UserService,
	private val successHandler: OAuth2LoginSuccessHandler,
	private val failureHandler: OAuth2LoginFailureHandler
) {

	@Value("\${next.server.url}")
	private lateinit var frontUrl: String

	@Bean
	fun securityFilterChain(http: HttpSecurity): SecurityFilterChain {
		http
			.cors { it.configurationSource(corsConfigurationSource()) }
			.csrf(AbstractHttpConfigurer<*, *>::disable)
			.headers { it.frameOptions { frameOptions -> frameOptions.disable() } }
			.sessionManagement { it.sessionCreationPolicy(SessionCreationPolicy.STATELESS) }
			.authorizeHttpRequests { auth ->
				auth
					.requestMatchers("/api/admin/**").hasAuthority("ROLE_ADMIN")
				    .anyRequest().permitAll()
			}
			.exceptionHandling { it.authenticationEntryPoint(customAuthenticationEntryPoint) }
			.addFilterBefore(jwtAuthenticationFilter, UsernamePasswordAuthenticationFilter::class.java)
			.oauth2Login { oauth2 ->
				oauth2
					.userInfoEndpoint { userInfo -> userInfo.userService(oAuth2UserService) }
					.successHandler(successHandler)
					.failureHandler(failureHandler)
			}

		return http.build()
	}

	@Bean
	fun corsConfigurationSource(): CorsConfigurationSource {
		val configuration = CorsConfiguration()
		configuration.allowedOriginPatterns = listOf(frontUrl)
		configuration.allowedMethods = listOf("GET", "POST", "PUT", "DELETE", "OPTIONS")
		configuration.allowedHeaders = listOf("*")
		configuration.allowCredentials = true
		configuration.addExposedHeader("Authorization")

		val source = UrlBasedCorsConfigurationSource()
		source.registerCorsConfiguration("/**", configuration)
		return source
	}

	@Bean
	fun passwordEncoder(): PasswordEncoder {
		return Argon2PasswordEncoder(16, 32, 1, 4096, 3)
	}
}
